packages/core/src/lib/callback-handler.ts

```diff
@@ -122,116 +124,203 @@ export async function handleLogin(
+        // create a new account for the user, link it to the OAuth account and
+        // create a new session for them so they are signed in with it.
+        const { id: _, ...newUser } = { ...profile, emailVerified: null }
+        user = await createUser(newUser)
+      }
+      await events.createUser?.({ user })
 
-    if (user) {
-      // If the user is already signed in and the OAuth account isn't already associated
-      // with another user account then we can go ahead and link the accounts safely.
       await linkAccount({ ...account, userId: user.id })
       await events.linkAccount?.({ user, account, profile })
 
-      // As they are already signed in, we don't need to do anything after linking them
-      return { session, user, isNewUser }
+      session = useJwtSession
+        ? {}
+        : await createSession({
+            sessionToken: generateSessionToken(),
+            userId: user.id,
+            expires: fromDate(options.session.maxAge),
+          })
+
+      return { session, user, isNewUser: true }
     }
+  } else if (account.type === "passkey") {
 
-    // If the user is not signed in and it looks like a new OAuth account then we
-    // check there also isn't an user account already associated with the same
-    // email address as the one in the OAuth profile.
-    //
-    // This step is often overlooked in OAuth implementations, but covers the following cases:
-    //
-    // 1. It makes it harder for someone to accidentally create two accounts.
-    //    e.g. by signin in with email, then again with an oauth account connected to the same email.
-    // 2. It makes it harder to hijack a user account using a 3rd party OAuth account.
-    //    e.g. by creating an oauth account then changing the email address associated with it.
-    //
-    // It's quite common for services to automatically link accounts in this case, but it's
-    // better practice to require the user to sign in *then* link accounts to be sure
-    // someone is not exploiting a problem with a third party OAuth service.
-    //
-    // OAuth providers should require email address verification to prevent this, but in
-    // practice that is not always the case; this helps protect against that.
-    const userByEmail = profile.email
-      ? await getUserByEmail(profile.email)
-      : null
-    if (userByEmail) {
-      const provider = options.provider as OAuthConfig<any>
-      if (provider?.allowDangerousEmailAccountLinking) {
-        // If you trust the oauth provider to correctly verify email addresses, you can opt-in to
-        // account linking even when the user is not signed-in.
-        user = userByEmail
+    // We want to make sure our adapter implements the required methods.
+    // Obs: we use ! here because we know from above that the adapter is not undefined.
+    const adapter: Adapter = options.adapter!
+
+    assertAdapterImplementsMethods(
```

Comment:
Same as comment as above, let's throw the error earlier when calling assert config, this would be more useful for the developers IMO

Reply:
Fixed in 0635795efddefd044c5d776fb1661127c80215b3